### ISSUE TITLE
`block-scoped-var` should check re-declarations and shadowing.

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#3158]
I thought `block-scoped-var` rule emulates the block-scoped bindings for `var`.

But currently, the rule does not warn re-declarations and shadowing. And the rule reports duplicate warnings of `no-undef` rule.

I want to add checking re-declarations and shadowing into `block-scoped-var` rule.
Also I want to remove checking duplicates of `no-undef` rule.

- `no-redeclare` rule does conflict with `block-scoped-var` rule. So `block-scoped-var` rule will generate duplicates of `no-redeclare` rule also.
- `no-shadow` rule does not conflict with `block-scoped-var` rule. So block-scoped-var rule will generate warnings about shadowing only in real re-declarations.

I'm happy to work on this if accepted.

## test.js

```js
/* eslint block-scoped-var: 2, no-shadow:2, no-redeclare: 2, no-undef:2 */
{
  var a = 0;
  if (true) {
    var a = 1; // shadowed `var a = 0`
  }
  else {
    var a = 2; // shadowed `var a = 0`
  }
  var a = 3; // redeclared with `var a = 0`

  function foo() {
    var a = 4; // shadowed all other `a`s.
  }
}
bar(a);
```

## Command & Actual

```
> eslint -v
v1.0.0-rc-2

> eslint test.js --no-eslintrc

test.js
   5:9  error  a is already defined                      no-redeclare
   8:9  error  a is already defined                      no-redeclare
  10:7  error  a is already defined                      no-redeclare
  13:9  error  a is already declared in the upper scope  no-shadow
  16:1  error  "bar" used outside of binding context     block-scoped-var
  16:1  error  "bar" is not defined                      no-undef
```

## Expected

```
test.js
   5:9  error  a is already defined                      no-redeclare
   5:9  error  a is already declared in the upper scope  block-scoped-var
   8:9  error  a is already defined                      no-redeclare
   8:9  error  a is already declared in the upper scope  block-scoped-var
  10:7  error  a is already defined                      no-redeclare
  10:7  error  a is already defined                      block-scoped-var
  13:9  error  a is already declared in the upper scope  no-shadow
  16:1  error  "bar" is not defined                      no-undef
  16:5  error  "a" is not defined                        block-scoped-var
```

When turned `no-redeclare` off:

```
test.js
   5:9  error  a is already declared in the upper scope  block-scoped-var
   8:9  error  a is already declared in the upper scope  block-scoped-var
  10:7  error  a is already defined                      block-scoped-var
  13:9  error  a is already declared in the upper scope  no-shadow
  16:1  error  "bar" is not defined                      no-undef
  16:5  error  "a" is not defined                        block-scoped-var
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
`block-scoped-var` only makes sure that you use a variable within the same block it was defined. It's not supposed to fully emulate block bindings. Does that change your thinking?

**Comment by mysticatea** [comment]
Ah, I'm sorry, I misunderstood from the rule name...

Please feel free to close this issue.
But this implement is not difficult, so I can make if anybody wanted it.
